doc/man/genesi-update.1.scd
genesi-update(1) ["Genesi-Update 3.19.4" ["Genesi-Update Manual"]]

# NAME

Genesi-Update - An interactive update notifier & applier for Arch Linux that assists you with important pre / post update tasks.

# SYNOPSIS

*genesi-update* [OPTION]

# DESCRIPTION

An interactive update notifier & applier for Arch Linux that assists you with important pre / post update tasks.
Includes a dynamic & clickeable systray applet for an easy integration with any Desktop Environment / Window Manager.

Genesi-Update is designed to follow usual system maintenance steps, as described in the Arch Wiki at https://wiki.archlinux.org/title/System_maintenance.

Features:

- Automatic check and listing of available updates.
- Check for recent Arch Linux news (and offers to display them if there are).
- Check for orphan packages (and offers to remove them if there are).
- Check for old & uninstalled packages in cache (and offers to remove them if there are).
- Check for pacnew & pacsave files (and offers to process them if there are).
- Check for pending kernel update requiring a reboot (and offers to do so if there's one).
- Check for services requiring a post upgrade restart (and offers to do so if there are).
- Support for `sudo`, `sudo-rs`, `doas` & `run0`.

Optional support for:

- AUR packages (via `paru`, `yay` or `pikaur`).
- Flatpak packages (via `flatpak`).
- Desktop notifications (via `libnotify`).
- Check for ALHP build queue or outdated mirrors (via `alhp-utils`).

# OPTIONS

If no option is passed, launch the relevant series of functions to perform a complete and proper system update (can also be triggered by clicking the systray applet).

*-c, --check*
	Check for available updates. If there are, send a desktop notification and update the systray icon.
	To run the `--check` option automatically and periodically, see the *"The systemd timer"* chapter in the *"USAGE"* section below.

*-l, --list*
	Display the list of pending updates.

*-d, --devel*
	Include AUR development packages updates.

*-n, --news*
	Display latest Arch news.
	You can optionally specify how much Arch news to display as an argument, like so: `genesi-update --news 10`. Defaults to 5.

*-s, --services*
	Check for services requiring a post upgrade restart.

*-D, --debug*
	Display debug traces.

*--gen-config*
	Generate a default / example `genesi-update.conf` configuration file (see the *genesi-update.conf*(5) man page for more details).
	You can optionally pass the `--force` argument to overwrite any existing `genesi-update.conf` configuration file.

*--show-config*
	Display the content of the `genesi-update.conf` configuration file (if it exists).

*--edit-config*
	Edit the `genesi-update.conf` configuration file (if it exists).

*--tray*
	Start the Genesi-Update systray applet.
	To start it automatically at boot, see the *"The systray applet"* chapter in the *"USAGE"* section below.

*-V, --version*
	Display version information.

*-h, --help*
	Display the help message.

Certain options can be enabled, disabled or modified via the `genesi-update.conf` configuration file, see the *genesi-update.conf*(5) man page for more details.

# USAGE

The usage consist of starting the systray applet and enabling the systemd timer.

## The systray applet

To start the systray applet, launch the *Genesi-Update Systray Applet* application from your app menu.

To start it automatically at boot, you can either:

- Run the following command (preferred method for most Desktop Environments, uses XDG Autostart): 

```
genesi-update --tray --enable
```

- Enable the associated systemd service (in case your Desktop Environment doesn't support XDG Autostart): 

```
systemctl --user enable --now genesi-update-tray.service
```

- Add the following command to your "auto-start" apps / configuration file (in case you use a Window Manager or a Wayland Compositor):

```
genesi-update --tray
```

*If the systray applet doesn't start at boot regardless or if it doesn't work as expected* (e.g. the icon is missing or the click actions do not act as they should), please read the *"The systray applet does not start at boot or does not work as expected"* chapter in the *"TIPS AND TRICKS"* section below.

The systray icon dynamically changes to indicate the current state of your system ("up to date" or "updates available"). When clicked, it launches `genesi-update` in a terminal window via the `genesi-update.desktop` file.

*If clicking the systray applet does nothing*, please read the *"Run Genesi-Update in a specific terminal emulator"* chapter in the *"TIPS AND TRICKS"* section below.

## The systemd timer

To perform automatic and periodic checks for available updates, enable the associated systemd timer:

```
systemctl --user enable --now genesi-update.timer
```

By default, a check is performed at *boot and then once every hour*. The check cycle can be customized, see the *"Modify the check cycle"* chapter in the *"TIPS AND TRICKS"* section below.

# TIPS AND TRICKS 

## The systray applet does not start at boot or does not work as expected

Make sure you followed instructions of the *"The systray applet"* chapter in the *"USAGE"* section above.

If the systray applet doesn't start at boot regardless or if it doesn't work as expected (e.g. the icon is missing or the click actions do not act as they should), this could be the result of a *race condition*.

To prevent that, you can add a small delay to the systray applet startup using the `sleep` command:

- If you used `genesi-update --tray --enable`, modify the `Exec=` line in the `genesi-update-tray.desktop` file (which is under `~/.config/autostart/` by default), like so: 

```
[...]
Exec=/bin/sh -c "sleep 3 && genesi-update --tray"
[...]
```

- If you used the `genesi-update-tray.service` systemd service, run `systemctl --user edit --full genesi-update-tray.service` and modify the `ExecStart=` line, like so:

```
[...]
ExecStart=/bin/sh -c "sleep 3 && genesi-update --tray"
[...]
```

- If you're using a standalone Window Manager or a Wayland Compositor, modify the command in your "auto-start" apps / your configuration file, like so:

```
[...]
sleep 3 && genesi-update --tray
[...]
```

If the systray applet still does not start at boot, try to gradually increase the `sleep` value.

## Modify the check cycle

If you enabled the systemd timer, a check for available updates is automatically launched at boot and then once a day.

If you want to customize the check cycle, run `systemctl --user edit --full genesi-update.timer` and modify the `OnUnitActiveSec` value to your liking.
For instance, if you want `Genesi-Update` to check for new updates every hour instead:

```
[...]
[Timer]
OnStartupSec=2min
RandomizedDelaySec=1h
OnUnitActiveSec=1h
[...]
```

Time units are *s* for seconds, *m* for minutes, *h* for hours, *d* for days...
See https://www.freedesktop.org/software/systemd/man/latest/systemd.time.html#Parsing%20Time%20Spans for more details.

In case you want `Genesi-Update` to check for new updates only once at boot, you can simply delete the `OnUnitActiveSec` line completely.

## Run Genesi-Update in a specific terminal emulator

"gio" (used to launch the `genesi-update` terminal application via the `genesi-update.desktop` file when the systray applet is clicked) currently has a default limited list of known terminal emulators.  
As such, if you don't have any of these "known" terminal emulators installed on your system, you might face an issue where clicking the systray applet does nothing (as `gio` couldn't find a terminal emulator from the said list). Incidentally, you might have multiple terminal emulators installed on your system. In both cases, you can specify which terminal emulator to use.

To do so, install the `xdg-terminal-exec` AUR package (https://aur.archlinux.org/packages/xdg-terminal-exec), create the `~/.config/xdg-terminals.list` file and add the name of the `.desktop` file of your terminal emulator of choice in it (e.g. `Alacritty.desktop`).

See https://github.com/Vladimir-csp/xdg-terminal-exec?tab=readme-ov-file#configuration for more details.

# EXIT STATUS

*0*
	OK.

*1*
	Invalid option.

*2*
	No privilege elevation command (`sudo`, `sudo-rs`, `doas` or `run0`) is installed or the one set in the `genesi-update.conf` configuration file isn't found.

*3*
	Error when launching the Genesi-Update systray applet.

*4*
	User didn't gave the confirmation to proceed.

*5*
	Error when updating packages.

*6*
	Error when calling the reboot command to apply a pending kernel update.

*7*
	No pending update when using the `-l / --list` option.

*8*
	Error when generating a configuration file with the `--gen-config` option.

*9*
	Error when reading the configuration file with the `--show-config` option.

*10*
	Error when creating the autostart desktop file for the systray applet with the `--tray --enable` option.

*11*
	Error when restarting services that require a post upgrade restart.

*12*
	Error during the pacnew files processing.

*13*
	Error when editing the configuration file with the `--edit-config` option.

*14*
	Libraries directory not found.

*15*
	The "diff prog" editor set in the `genesi-update.conf` configuration file isn't found.

*16*
	Error when creating the `statedir` and / or the `tmpdir` directories.

*17*
	Another instance of Genesi-Update is already running.

*18*
	Error when running Genesi-Update from the desktop notification action.

# SEE ALSO

*checkupdates*(8),
*pacman*(8),
*pacdiff*(8),
*paccache*(8),
*paru*(8),
*yay*(8),
*pikaur*(1),
*flatpak*(1),
*genesi-update.conf*(5)

# BUGS

Please report bugs to the GitHub page: https://github.com/Antiz96/genesi-update/issues

# AUTHOR

Robin Candau <robincandau@protonmail.com>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [v3.19.4](https://github.com/Antiz96/genesi-update/releases/tag/v3.19.4) - 2026-05-07

### Features

- *(i18n)* Add Belarusian translation ([#594](https://github.com/Antiz96/genesi-update/pull/594)) - ([3937ee9](https://github.com/Antiz96/genesi-update/commit/3937ee9c4b4ffbc4e512adc7c4e5f314397fd7a6)) by @lasaczka

## [v3.19.3](https://github.com/Antiz96/genesi-update/releases/tag/v3.19.3) - 2026-04-26

### Fixes

- *(orphans)* Additionally check for orphans packages before AUR package updates ([#587](https://github.com/Antiz96/genesi-update/pull/587)) - ([af74954](https://github.com/Antiz96/genesi-update/commit/af7495449f56c36c9344ca287829918a2b9de391)) by @Antiz96
- *(restart services)* Move condition logic outside of the library script ([#589](https://github.com/Antiz96/genesi-update/pull/589)) - ([2c42319](https://github.com/Antiz96/genesi-update/commit/2c42319a7b8bafcb9f9e604df67885cff635b6e6)) by @Antiz96
- *(state)* Update global update state files after each update processes ([#590](https://github.com/Antiz96/genesi-update/pull/590)) - ([8d3ee00](https://github.com/Antiz96/genesi-update/commit/8d3ee0098257e69434fad216bae16924113073d7)) by @Antiz96
- Update state files updating logic to match changes added in #583 ([#584](https://github.com/Antiz96/genesi-update/pull/584)) - ([3fe9e14](https://github.com/Antiz96/genesi-update/commit/3fe9e14735d6dff8b76649b59167f9f4c9a94fae)) by @Antiz96
- Treat issues during AUR and Flatpak packages updates as warning ([#583](https://github.com/Antiz96/genesi-update/pull/583)) - ([bf5c825](https://github.com/Antiz96/genesi-update/commit/bf5c825c849050688ddc5adee733549e115ea11f)) by @Antiz96

### Styling

- Add missing line break in Flatpak warning message ([#586](https://github.com/Antiz96/genesi-update/pull/586)) - ([ba08c8e](https://github.com/Antiz96/genesi-update/commit/ba08c8e612e7d9df9136e76b77b986c570e68bd8)) by @Antiz96

### Miscellaneous

- *(i18n)* Update translation files ([#591](https://github.com/Antiz96/genesi-update/pull/591)) - ([32b56bf](https://github.com/Antiz96/genesi-update/commit/32b56bf4abf2ac12dedc53af4bd6754a9414b07f)) by @Antiz96
- *(icon)* Remove useless update calls ([#585](https://github.com/Antiz96/genesi-update/pull/585)) - ([ca5a3af](https://github.com/Antiz96/genesi-update/commit/ca5a3af58dbf8c499e2f565d044711da1befac30)) by @Antiz96
- Add execution bit to the flatpak_unused_packages library script ([#588](https://github.com/Antiz96/genesi-update/pull/588)) - ([ceefadb](https://github.com/Antiz96/genesi-update/commit/ceefadbbcac87e4c542e862f5999f0983989b330)) by @Antiz96

## [v3.19.2](https://github.com/Antiz96/genesi-update/releases/tag/v3.19.2) - 2026-04-14

### Fixes

- *(flatpak)* Fix interactive polkit authentications handling in TTY / SSH environment ([#578](https://github.com/Antiz96/genesi-update/pull/578)) - ([7d3e893](https://github.com/Antiz96/genesi-update/commit/7d3e893c321ecb5ab21a2456c5c9723f53314f28)) by @FineWolf

### Miscellaneous

- *(i18n)* Update translation files ([#580](https://github.com/Antiz96/genesi-update/pull/580)) - ([524828f](https://github.com/Antiz96/genesi-update/commit/524828ff7cf6b80ff86294d5ca60c2d95116d0f5)) by @Antiz96
- Update contributing guidelines regarding derivatives ([#579](https://github.com/Antiz96/genesi-update/pull/579)) - ([2fa6720](https://github.com/Antiz96/genesi-update/commit/2fa67207b2ab4d493b539478e8c0d597baf87287)) by @Antiz96

## [v3.19.1](https://github.com/Antiz96/genesi-update/releases/tag/v3.19.1) - 2026-04-05

### Fixes

- Syntax fix in zsh completion ([#572](https://github.com/Antiz96/genesi-update/pull/572)) - ([00c86ef](https://github.com/Antiz96/genesi-update/commit/00c86eff71a221f7064ce3d9dcaf19d359ccf9a4)) by @balki

### Miscellaneous

- *(i18n)* Update Brazilian Portuguese and Spanish translation ([#571](https://github.com/Antiz96/genesi-update/pull/571)) - ([d857385](https://github.com/Antiz96/genesi-update/commit/d857385bdf2ed3d62c8272b96e746a51da378c65)) by @diegons490

## [v3.19.0](https://github.com/Antiz96/genesi-update/releases/tag/v3.19.0) - 2026-03-27

### Features

- *(i18n)* Add Basque translation ([#545](https://github.com/Antiz96/genesi-update/pull/545)) - ([93c2663](https://github.com/Antiz96/genesi-update/commit/93c26635c276c5203a1feaafa6b2906170bb0dd8)) by @bingenm
- Add a default timeout for update check requests ([#554](https://github.com/Antiz96/genesi-update/pull/554)) - ([28d7d4e](https://github.com/Antiz96/genesi-update/commit/28d7d4ebdccc93c8c7cb733be3c4c76ad5dc692a)) by @Antiz96
- Disable flatpak support if no flatpak package is installed ([#552](https://github.com/Antiz96/genesi-update/pull/552)) - ([e45185a](https://github.com/Antiz96/genesi-update/commit/e45185a70b0747530be0d7a0bb43a784cb405af4)) by @Antiz96
- Add support for ALHP utils package check ([#539](https://github.com/Antiz96/genesi-update/pull/539)) - ([30551c4](https://github.com/Antiz96/genesi-update/commit/30551c49f883e861bc86d65df22922d021ae96ca)) by @kescherCode

### Styling

- Add line breaks in alhp check warning messages ([#555](https://github.com/Antiz96/genesi-update/pull/555)) - ([a215705](https://github.com/Antiz96/genesi-update/commit/a2157054c4bb0266ea710f2050a479bb77160483)) by @Antiz96
- Formatting fixes ([#540](https://github.com/Antiz96/genesi-update/pull/540)) - ([c9eb9fd](https://github.com/Antiz96/genesi-update/commit/c9eb9fd6a1ac6f395d6fed3a0b254effcab77650)) by @Antiz96

### Miscellaneous

- *(i18n)* Update Norwegian translation ([#568](https://github.com/Antiz96/genesi-update/pull/568)) - ([0693b93](https://github.com/Antiz96/genesi-update/commit/0693b93b30189c5015d4e136f5237229a389839d)) by @Kjodleiken
- *(i18n)* Update Russian translation ([#567](https://github.com/Antiz96/genesi-update/pull/567)) - ([efe973a](https://github.com/Antiz96/genesi-update/commit/efe973a7509396a58d7b748fdd6cb875a7b40607)) by @DrWeb56

src/genesi-update.sh
#!/bin/bash

# genesi-update: An interactive update notifier & applier for Arch Linux that assists you with important pre / post update tasks
# https://github.com/Antiz96/genesi-update
# SPDX-License-Identifier: GPL-3.0-or-later

# General variables
name="genesi-update"
_name="Genesi-Update"
version="3.19.4"
option="${1}"

# Define the directory containing libraries
if [ -n "${ARCH_UPDATE_LIBDIR}" ]; then
	libdir="${ARCH_UPDATE_LIBDIR}"
elif [ -d "${XDG_DATA_HOME}/${name}/lib" ]; then
	libdir="${XDG_DATA_HOME}/${name}/lib"
elif [ -d "${HOME}/.local/share/${name}/lib" ]; then
	libdir="${HOME}/.local/share/${name}/lib"
elif [ -d "${XDG_DATA_DIRS}/${name}/lib" ]; then
	libdir="${XDG_DATA_DIRS}/${name}/lib"
elif [ -d "/usr/local/share/${name}/lib" ]; then
	libdir="/usr/local/share/${name}/lib"
elif [ -d "/usr/share/${name}/lib" ]; then
	libdir="/usr/share/${name}/lib"
else
	echo -e >&2 "ERROR: Libraries directory not found"
	exit 14
fi

# Source the "config" library which checks options set in the genesi-update.conf configuration file
# shellcheck source=src/lib/config.sh
source "${libdir}/config.sh"

# Source the "common" library which sets variables, functions and parameters commonly used across the various Genesi-Update stages
# shellcheck source=src/lib/common.sh
source "${libdir}/common.sh"

# Source the different libraries depending on the option(s) passed
case "${option}" in
	"")
		# Check for requirements
		check_su_cmd
		check_aur_helper
		check_diff_prog

		# Source the "full_upgrade" library which executes the series of relevant functions / libraries to perform a complete and proper update of the system
		# shellcheck source=src/lib/full_upgrade.sh
		source "${libdir}/full_upgrade.sh"
	;;
	-d|--devel)
		# Set the "--devel" flag for AUR helpers
		devel_flag+=("--devel")

		# Check for requirements
		check_su_cmd
		check_aur_helper
		check_diff_prog

		# Source the "full_upgrade" library which executes the series of relevant functions / libraries to perform a complete and proper update of the system
		# shellcheck source=src/lib/full_upgrade.sh
		source "${libdir}/full_upgrade.sh"
	;;
	-c|--check)
		# Check for requirements
		check_aur_helper

		# Source the "check" library which checks for available updates
		# shellcheck source=src/lib/check.sh
		source "${libdir}/check.sh"
	;;
	-l|--list)
		# Check for requirements
		check_aur_helper

		# Set that the "-l / --list" option is used (required for the "list_packages" library to only print the list of packages available for update without asking for the user confirmation to apply them)
		list_option="true"

		# Source the "list_packages" library which displays the list of packages available for updates
		# shellcheck source=src/lib/list_packages.sh
		source "${libdir}/list_packages.sh"
	;;
	-n|--news)
		# Set that news should be shown and that the "-n / --news" option is used (required for the "list_news" library to only print the list of recent Arch news without looking to continue with the update process)
		show_news="true"
		news_option="true"

		# Check if the user specified a specific number of news to display
		if [ "${2}" -gt 0 ] 2> /dev/null; then
			news_num="${2}"
		fi

		# Source the "list_news" library which displays the latest Arch news and offers to read them
		# shellcheck source=src/lib/list_news.sh
		source "${libdir}/list_news.sh"
	;;
	-s|--services)
		# Check for requirements
		check_su_cmd

		# Source the "restart_services" library which displays services requiring a post upgrade restart (and offers to do so if there are)
		# shellcheck source=src/lib/restart_services.sh
		source "${libdir}/restart_services.sh"
	;;
	--gen-config)
		# Check if the user specified to overwrite any existing "genesi-update.conf" configuration file
		if [ "${2}" == "--force" ]; then
			overwrite_config_file="true"
		fi

		# Source the "gen-config" library which generates a default "genesi-update.conf" configuration file (if it doesn't exists yet)
		# shellcheck source=src/lib/gen-config.sh
		source "${libdir}/gen-config.sh"
	;;
	--show-config)
		# Source the "show-config" library which displays the current "genesi-update.conf" configuration file
		# shellcheck source=src/lib/show-config.sh
		source "${libdir}/show-config.sh"
	;;
	--edit-config)
		# Source the "edit-config" library which edits the current "genesi-update.conf" configuration file
		# shellcheck source=src/lib/edit-config.sh
		source "${libdir}/edit-config.sh"
	;;
	--tray)
		# Source the "tray" library which starts the Genesi-Update systray applet
		# shellcheck source=src/lib/tray.sh
		source "${libdir}/tray.sh"
	;;
	-h|--help)
		# Source the "help" library which displays the help message
		# shellcheck source=src/lib/help.sh
		source "${libdir}/help.sh"
	;;
	-V|--version)
		# Source the "version" library which displays version information
		# shellcheck source=src/lib/version.sh
		source "${libdir}/version.sh"
	;;
	*)
		# Source the "invalid_option" library which displays an error when an invalid option is passed
		# shellcheck source=src/lib/invalid_option.sh
		source "${libdir}/invalid_option.sh"
	;;
esac
